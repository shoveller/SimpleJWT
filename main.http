POST http://0.0.0.0:8000/login
Content-Type: application/json

{
  "username": "땡탄이"
}
