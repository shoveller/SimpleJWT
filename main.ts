import {create, getNumericDate, verify} from "djwt";

const alg = "HS512";
const key = await crypto.subtle.generateKey(
    { name: "HMAC", hash: "SHA-512" },
    true,
    ["sign", "verify"],
);

type PayloadType = {
    iss: string;
    sub: string;
    exp: number;
};

// 토큰을 생성한다
export async function generateJwt(payload: PayloadType) {
    return await create({ alg, typ: "JWT" }, payload, key);
}

export async function getAccessToken(username: string) {
    return await generateJwt({
        iss: "your-app",
        sub: username,
        exp: getNumericDate(60 * 60), // 1시간 유효
    });
}

// 리프레시 토큰 생성. 리프레시 토큰은 유효기간이 길다. 나머지는 액세스 토큰과 같다
export async function getRefreshToken(username: string) {
    return await generateJwt({
        iss: "your-app",
        sub: username,
        exp: getNumericDate(60 * 60 * 24 * 7), // 1주일 유효
    });
}

// JWT 토큰 검증
export async function verifyJwt(token: string) {
    try {
        return await verify(token, key);
    } catch (error) {
        console.error("Invalid JWT:", error);
        return null;
    }
}

// 로그인
async function login(req: Request) {
    const { username } = await req.json();
    if (!username) {
        return new Response("Username is required", { status: 400 });
    }

    // Access Token과 Refresh Token 생성
    const accessToken = await getAccessToken(username);
    const refreshToken = await getRefreshToken(username);

    return new Response(JSON.stringify({ accessToken, refreshToken }), {
        status: 200,
        headers: { "Content-Type": "application/json" },
    });
}

// 보호된 경로 핸들러
async function resource(req: Request) {
    const authHeader = req.headers.get("Authorization");
    if (!authHeader || !authHeader.startsWith("Bearer ")) {
        // 토큰이 없으면 인증 실패
        return new Response("Unauthorized", { status: 401 });
    }

    const token = authHeader.substring(7);
    const payload = await verifyJwt(token);
    if (payload) {
        return new Response(`Hello, ${payload.sub}!`, { status: 200 });
    }

    // 토큰이 검증을 통과하지 못하면 토큰 실패
    return new Response("Invalid token", { status: 401 });
}

// 토큰 갱신 핸들러 (리프레시 토큰 사용)
async function tokenRefresh(req: Request) {
    const { refreshToken } = await req.json();
    // 리프레시 토큰이 없으면 400 처리
    if (!refreshToken) {
        return new Response("Refresh token is required", { status: 400 });
    }

    // 리프레시 토큰 검증에 실패하면 401 처리
    const payload = await verifyJwt(refreshToken);
    if (!payload) {
        return new Response("Invalid refresh token", { status: 401 });
    }

    // 토큰 검증에 성공하면 새로운 토큰 발급
    const accessToken = await getAccessToken(payload.sub!);
    return new Response(JSON.stringify({ accessToken }), {
        status: 200,
        headers: { "Content-Type": "application/json" },
    });
}

if (import.meta.main) {
    Deno.serve((req) => {
        const url = new URL(req.url);
        if (url.pathname === "/") {
            return resource(req);
        }
        if (url.pathname === "/login" && req.method === "POST") {
            return login(req);
        }
        if (url.pathname === "/refresh" && req.method === "POST") {
            return tokenRefresh(req);
        }

        return new Response("404", { status: 404 });
    });
}
