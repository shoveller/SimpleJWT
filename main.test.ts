import {getRefreshToken, getAccessToken, verifyJwt,} from "./main.ts";
import { assertSnapshot } from "@std/testing/snapshot";
import { assertEquals } from '@std/assert'

Deno.test("액세스 토큰을 생성할 수 있다.", async (ctx) => {
    const jwt = await getAccessToken("김땡탄");
    await assertSnapshot(ctx, jwt)
})

Deno.test("리프레시 토큰을 생성할 수 있다.", async (ctx) => {
    const jwt = await getRefreshToken("김땡탄");
    await assertSnapshot(ctx, jwt)
})

Deno.test("액세스 토큰을 검증할 수 있다.", async (ctx) => {
    const jwt = await getAccessToken("김땡탄");
    const payload = await verifyJwt(jwt)
    await assertSnapshot(ctx, payload);

    const fakePayload = await verifyJwt('')
    assertEquals(fakePayload, null);
    await assertSnapshot(ctx, fakePayload);
})
