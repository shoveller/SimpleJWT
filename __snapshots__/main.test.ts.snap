export const snapshot = {};

snapshot[`액세스 토큰을 생성할 수 있다. 1`] = `"eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJ5b3VyLWFwcCIsInN1YiI6Iuq5gOuVoe2DhCIsImV4cCI6MTcyMzI4MTExN30.sAwE7ck1x3NS_AQ7e-uc-WKyzOPURjPlNftLzjguIN4ZrOUpQ4LhNsOWbMqDf7EicsG25JCj2BLlFQ35KROhww"`;

snapshot[`리프레시 토큰을 생성할 수 있다. 1`] = `"eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJ5b3VyLWFwcCIsInN1YiI6Iuq5gOuVoe2DhCIsImV4cCI6MTcyMzg4MjMxN30.NIRg4zcmW7TsbcH2xV1sf008fWoOxT-0jkXPt2W_P8g3S8Tj7csFR8g-9M9i-g-gwGmFDhahecaf7oku6dPJzg"`;

snapshot[`액세스 토큰을 검증할 수 있다. 1`] = `
{
  exp: 1723281117,
  iss: "your-app",
  sub: "김땡탄",
}
`;

snapshot[`액세스 토큰을 검증할 수 있다. 2`] = `null`;
